src/Interpreters/ArithmeticOperationsInAgrFuncOptimize.cpp

```diff
@@ -0,0 +1,316 @@
+        for (const auto &ar: it.first)
+        {
+            auto num = ar->as<ASTLiteral>()->value.get<Int128>();
+
+            /// if multiplication is negative, min function becomes max
+
+            if ((ar->as<ASTLiteral>()->value.getType() == Field::Types::Int64 ||
+                 ar->as<ASTLiteral>()->value.getType() == Field::Types::Int128) && static_cast<int64_t>(num) < 0)
+                tp *= -1;
+        }
+
+        std::pair<int, std::string> cur = {tp, max};
+        buildTree(f_n, min, mul, it, cur);
+    }
+    else if (inter_node && inter_node->name == plus)
+    {
+        std::pair<ASTs, ASTs> it = findAllConsts(f_n, plus);
+        std::pair<int, std::string> cur = {1, min};
+        buildTree(f_n, min, plus, it, cur);
+    }
+    else
+    {
+        return;
+    }
+}
+
+void maxOptimize(ASTFunction * f_n)
+{
+    std::string max = "max";
+    std::string min = "min";
+    std::string mul = "multiply";
+    std::string plus = "plus";
+
+    const auto * function_args = f_n->arguments->as<ASTExpressionList>();
+
+    if (!function_args || function_args->children.size() != 1)
+        throw Exception("Wrong number of arguments for function 'max' (" + toString(function_args->children.size()) + " instead of 1)",
+                        ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+    auto * inter_node = f_n->arguments->children[0]->as<ASTFunction>();
+    if (inter_node && inter_node->name == mul)
+    {
+        int tp = 1;
+        std::pair<ASTs, ASTs> it = findAllConsts(f_n, mul);
+
+        if (it.first.empty())
+            return;
+
+        for (const auto &ar: it.first)
+        {
+            auto num = ar->as<ASTLiteral>()->value.get<Int128>();
+
+            /// if multiplication is negative, max function becomes min
+
+            if ((ar->as<ASTLiteral>()->value.getType() == Field::Types::Int64 ||
+                    ar->as<ASTLiteral>()->value.getType() == Field::Types::Int128) && static_cast<int64_t>(num) < 0)
+                tp *= -1;
+        }
+
+        std::pair<int, std::string> cur = {tp, min};
+        buildTree(f_n, max, mul, it, cur);
+    }
+    else if (inter_node && inter_node->name == plus)
+    {
+        std::pair<ASTs, ASTs> it = findAllConsts(f_n, plus);
+        std::pair<int, std::string> cur = {1, max};
+        buildTree(f_n, max, plus, it, cur);
+    }
+    else
+    {
+        return;
+    }
+}
+
+/// optimize for min, max, sum is ready, ToDo: any, anyLast, groupBitAnd, groupBitOr, groupBitXor
+void ArithmeticOperationsInAgrFuncVisitor::visit(ASTPtr & current_ast)
+{
+    if (!current_ast)
+        return;
+
+    for (ASTPtr & child : current_ast->children)
+    {
+        auto * function_node = child->as<ASTFunction>();
+
+        if (!function_node || !(function_node->name == "sum" || function_node->name == "min" || function_node->name == "max"))
+        {
+            visit(child);
+            continue;
+        }
+
+        if (function_node->name == "sum")
+        {
```

Comment:
remove all the branches. There're short conditions and short sunctions. Code is clearer without branches.